File: app/src/main/java/com/yao/zhihudaily/ui/daily/CommentAdapter.java
package com.yao.zhihudaily.ui.daily;

import android.content.Context;
import android.graphics.Bitmap;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.BitmapImageViewTarget;
import com.yao.zhihudaily.R;
import com.yao.zhihudaily.model.Comment;
import com.yao.zhihudaily.tool.GlideCircleTransform;

import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.graphics.drawable.RoundedBitmapDrawable;
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;

/**
 *
 * @author Administrator
 * @date 2016/8/30
 */
public class CommentAdapter extends RecyclerView.Adapter<CommentAdapter.ShortCommentViewHolder> {

    private static final int COMMENT = 0;
    private static final int COMMENT_WITH_REPLY = 1;


    private ArrayList<Comment> mComments = new ArrayList<>();
    private LayoutInflater mLayoutInflater;
    private Context ctx;
    private GlideCircleTransform mGlideCircleTransform;

    public CommentAdapter(Context ctx) {
        this.ctx = ctx;
        mLayoutInflater = LayoutInflater.from(ctx);
        mGlideCircleTransform = new GlideCircleTransform();
    }

    public void addList(ArrayList<Comment> comments) {
        this.mComments.addAll(comments);
    }


    @NonNull
    @Override
    public CommentAdapter.ShortCommentViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == COMMENT) {
            return new ShortCommentViewHolder(mLayoutInflater.inflate(R.layout.item_short_comment, null), COMMENT);
        } else {
            return new ShortCommentViewHolder(mLayoutInflater.inflate(R.layout.item_short_comment_with_reply, null), COMMENT_WITH_REPLY);
        }
    }

    @Override
    public void onBindViewHolder(final CommentAdapter.ShortCommentViewHolder holder, int position) {
        Comment c = mComments.get(position);
        BitmapImageViewTarget target = new BitmapImageViewTarget(holder.ivAvatar){
            @Override
            protected void setResource(Bitmap resource) {
                RoundedBitmapDrawable circularBitmapDrawable = RoundedBitmapDrawableFactory.create(ctx.getResources(), resource);
                circularBitmapDrawable.setCircular(true);
                holder.ivAvatar.setImageDrawable(circularBitmapDrawable);
            }
        };
        Glide.with(ctx).asBitmap().load(c.getAvatar()).centerCrop().into(target);
        Glide.with(ctx).load(c.getAvatar()).transform(mGlideCircleTransform).into(holder.ivAvatar);
        holder.tvAuthor.setText(c.getAuthor());
        holder.tvContent.setText(c.getContent());
        holder.tvTime.setText(c.getTimeStr());
        holder.tvLikes.setText(String.valueOf(c.getLikes()));
        if (c.getReplyTo() != null) {
            String replyAuthor = "@" + c.getReplyTo().getAuthor()+": ";
            if (holder.tvReplyAuthor != null) {
                holder.tvReplyAuthor.setText(replyAuthor);
            }
            if (holder.tvReplyContent != null) {
                holder.tvReplyContent.setText(replyAuthor + c.getReplyTo().getContent());
            }
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (mComments.get(position).getReplyTo() == null) {
            return COMMENT;
        } else {
            return COMMENT_WITH_REPLY;
        }
    }

    @Override
    public int getItemCount() {
        return mComments.size();
    }

    class ShortCommentViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.iv_avatar)
        ImageView ivAvatar;
        @BindView(R.id.tv_author)
        TextView tvAuthor;
        @BindView(R.id.tvContent)
        TextView tvContent;
        @BindView(R.id.tvTime)
        TextView tvTime;
        @BindView(R.id.tvLikes)
        TextView tvLikes;
        @Nullable
        @BindView(R.id.tvReplyAuthor)
        TextView tvReplyAuthor;
        @Nullable
        @BindView(R.id.tvReplyContent)
        TextView tvReplyContent;

        public ShortCommentViewHolder(View itemView, int type) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/yao/zhihudaily/ui/daily/CommentAdapter.kt
package com.yao.zhihudaily.ui.daily

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.CircleCrop
import com.bumptech.glide.request.target.BitmapImageViewTarget
import com.yao.zhihudaily.R
import com.yao.zhihudaily.model.Comment
import kotlinx.android.synthetic.main.item_short_comment_with_reply.view.*
import java.util.*

/**
 *
 * @author Yao
 * @date 2016/8/30
 */
class CommentAdapter(private val ctx: Context) : RecyclerView.Adapter<CommentAdapter.ShortCommentViewHolder>() {


    private val mComments = ArrayList<Comment>()
    private val mLayoutInflater: LayoutInflater
    private val mCircleCrop: CircleCrop

    init {
        mLayoutInflater = LayoutInflater.from(ctx)
        mCircleCrop = CircleCrop()
    }

    fun addList(comments: ArrayList<Comment>) {
        this.mComments.addAll(comments)
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ShortCommentViewHolder {
        return if (viewType == COMMENT) {
            ShortCommentViewHolder(mLayoutInflater.inflate(R.layout.item_short_comment, null), COMMENT)
        } else {
            ShortCommentViewHolder(mLayoutInflater.inflate(R.layout.item_short_comment_with_reply, null), COMMENT_WITH_REPLY)
        }
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: ShortCommentViewHolder, position: Int) {
        val c = mComments[position]
        val target = object : BitmapImageViewTarget(holder.ivAvatar) {
            override fun setResource(resource: Bitmap?) {
                val circularBitmapDrawable = RoundedBitmapDrawableFactory.create(ctx.resources, resource)
                circularBitmapDrawable.isCircular = true
                holder.ivAvatar.setImageDrawable(circularBitmapDrawable)
            }
        }
        Glide.with(ctx).asBitmap().load(c.avatar).centerCrop().into<BitmapImageViewTarget>(target)
        Glide.with(ctx).load(c.avatar).transform(mCircleCrop).into(holder.ivAvatar)
        holder.tvAuthor.text = c.author
        holder.tvContent.text = c.content
        holder.tvTime.text = c.timeStr
        holder.tvLikes.text = c.likes.toString()
        if (c.replyTo != null) {
            val replyAuthor = "@" + c.replyTo?.author + ": "
            holder.tvReplyAuthor.text = replyAuthor
            holder.tvReplyContent.text = replyAuthor + c.replyTo?.content
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (mComments[position].replyTo == null) {
            COMMENT
        } else {
            COMMENT_WITH_REPLY
        }
    }

    override fun getItemCount(): Int {
        return mComments.size
    }

    inner class ShortCommentViewHolder(view: View, type: Int) : RecyclerView.ViewHolder(view) {

        val ivAvatar: ImageView = view.iv_avatar
        val tvAuthor: TextView = view.tv_author
        val tvContent: TextView = view.tvContent
        val tvTime: TextView = view.tvTime
        val tvLikes: TextView = view.tvLikes
        val tvReplyAuthor: TextView = view.tvReplyAuthor
        val tvReplyContent: TextView = view.tvReplyContent
    }

    companion object {

        private val COMMENT = 0
        private val COMMENT_WITH_REPLY = 1
    }
}
